[AutoDiff] Improve @differentiable attribute type-checking.

dan-zheng opened the pull request:
Rewrite custom primal/adjoint func decl lookup function using
`TypeChecker.lookupUnqualified`.

The new lookup function is significantly simpler and more general. It handles overloaded decl refs, which the previous function didn't.

Fix various type-checking bugs, update tests.

dan-zheng committed:
[AutoDiff] Improve @differentiable attribute type-checking.

Rewrite custom primal/adjoint func decl lookup function using
`TC.lookupUnqualified`. The new lookup function is significantly simpler
and more general.

Fix various type-checking bugs, update tests.

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2214,124 +2214,70 @@ void AttributeChecker::visitImplementsAttr(ImplementsAttr *attr) {
+    }
+    else if (overloadNotFound) {
+      TC.diagnose(funcNameLoc, overloadDiagnostic);
+    } else {
```
Wrap this `else` because you wrapped the previous ones.

dan-zheng replied:
I actually unwrapped all the `if else`/`else`s because I prefer that.
Things are uniform now.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2549,26 +2524,40 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
+  auto adjointTypeCtx = original->getInnermostTypeContext();
+  if (!adjointTypeCtx) adjointTypeCtx = original->getParent();
+
+  Diagnostic adjointOverloadDiagnostic(
```
As we discussed in person, you can turn these into thunks.

dan-zheng replied:
Done.

dan-zheng committed:
[AutoDiff] Minor edits.

Convert diagnostics to thunks for efficiency, as requested by @rxwei.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
[AutoDiff] Use consistent name for `hasValidTypeContext` function.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2214,124 +2214,66 @@ void AttributeChecker::visitImplementsAttr(ImplementsAttr *attr) {
+  // Perform lookup.
+  auto results = TC.lookupUnqualified(lookupContext, funcName, funcNameLoc);
+  for (auto choice : results) {
+    if (auto decl = choice.getValueDecl()) {
```
Early return.

dan-zheng replied:
Done.

dan-zheng committed:
[AutoDiff] Minor edit.

Use early continue in for loop.

dan-zheng committed:
[AutoDiff] Add workaround for Swift 3 unqualified lookup behavior.

In Swift 3, static methods are not found via unqualified lookups:
https://github.com/apple/swift/commit/fbdecebf771a872582f1fd62088c558c2ea5fbf7

This is a problem because the stdlib is compiled with `-swift-version 3` for
Swift 3 compatibility and floating point types use the `@differentiable`
attribute with static adjoint methods like `_adjointAdd`.
